src/sidebar/util/test/wrap-react-component-test.js

```diff
@@ -0,0 +1,238 @@
+'use strict';
+
+const angular = require('angular');
+const { Component, createElement } = require('preact');
+const propTypes = require('prop-types');
+
+const wrapReactComponent = require('../wrap-react-component');
+const { createDirective } = require('../../directive/test/util');
+
+// Saved `onDblClick` prop from last render of `Button`.
+// This makes it easy to call it with different arguments.
+let lastOnDblClickCallback;
+
+function Button({ label, isDisabled, onClick, theme, onDblClick }) {
+  // We don't actually use the `onDblClick` handler in this component.
+  // It exists just to test callbacks that take arguments.
+  lastOnDblClickCallback = onDblClick;
+
+  return (
+    <button disabled={isDisabled} onClick={onClick} className={'btn--' + theme}>
+      {label}
+    </button>
+  );
+}
+Button.propTypes = {
+  // Simple input properties passed by parent component.
+  label: propTypes.string.isRequired,
+  isDisabled: propTypes.bool,
+
+  // A required callback with no arguments.
+  onClick: propTypes.func.isRequired,
+
+  // An optional callback with a `{ click }` argument.
+  onDblClick: propTypes.func,
+
+  // A property whose value comes from Angular dependency injection rather than
+  // the parent component.
+  theme: propTypes.string,
+};
+Button.injectedProps = ['theme'];
+
+describe('wrapReactComponent', () => {
+  function renderButton() {
+    const onClick = sinon.stub();
+    const onDblClick = sinon.stub();
+
+    const element = createDirective(document, 'btn', {
+      label: 'Edit',
+      isDisabled: false,
+      onClick,
+      onDblClick: {
+        args: ['count'],
+        callback: onDblClick,
+      },
+    });
+
+    return { element, onClick, onDblClick };
+  }
+
+  beforeEach(() => {
+    angular
+      .module('app', [])
+      .component('btn', wrapReactComponent(Button))
+      .value('theme', 'dark');
+    angular.mock.module('app');
+  });
+
+  afterEach(() => {
+    if (console.error.restore) {
+      console.error.restore();
+    }
+  });
+
+  it('derives Angular component "bindings" from React "propTypes"', () => {
+    const ngComponent = wrapReactComponent(Button);
+    assert.deepEqual(ngComponent.bindings, {
+      label: '<',
+      isDisabled: '<',
+      onClick: '&',
+      onDblClick: '&',
+
+      // nb. Props passed via dependency injection should not appear here.
+    });
+  });
+
+  it('renders the React component when the Angular component is created', () => {
+    const { element, onClick } = renderButton();
+
+    const btnEl = element[0].querySelector('button');
+    assert.ok(btnEl);
+
+    // Check that properties are passed correctly.
+    assert.equal(btnEl.textContent, 'Edit');
+    assert.equal(btnEl.disabled, false);
+
+    // Verify that events result in callbacks being invoked.
+    btnEl.click();
+    assert.called(onClick);
+  });
+
+  it('gets values for injected properties from Angular', () => {
```

Comment:
In reading the contents of this test, it's not immediately clear to me, from the outside world, how this tests properties from Angular...is it because of the `theme` value? Maybe a comment would help.

Reply:
Yes, exactly. The ‘theme’ value is marked as a dependency of the generated controller class for the component.